db and models setup

lugenx opened the pull request:
- connected to the database
- created mongoose models
- added routes
- created controller functions

some functions like user login and signup will be changed/improved when auth is applied.

mannan0819 commented on server/src/controller/userController.js:
```diff
@@ -0,0 +1,27 @@
+      name,
+      email,
+      postalCode,
+      password,
```
We prob want to hash the users password and not want to store the exact string. 
I think possible Argon2 is whats good to use right now
https://www.npmjs.com/package/argon2

lugenx replied:
yeah, agree. we will hash it with argon2 or bcrypt, it will be encrypted together with salt.  

i think we can work on encryption when we work on user auth, which is another issue i have created. just tried to reduce the size of prs and focus on one at a time. but that's really good point and it's so critical we shouldn't miss.

how does the database structure look like to you guys, i tried but still i feel like maybe there should be more efficient way to do it? 🤔 . 

we will probably change it anyway while we progress but i would like to try to make it as close as possible.

let me know if you guys have any questions or suggestions while examine this.

lugenx replied:
also need to do some research about which one is better to choose, argon2 or Bcrypt.

AbhiBajpaishek replied:
> We should also consider generating a random unique id string to each user and use that to identify them throughout the application as many apps do, as that will allow us to deal with situations in the future where it may be necessary to change a user's name, email or other information.

Mongo DB by default adds a unique id

mannan0819 reviewed (commented):
Just one comment to hash users passwords

lugenx commented:
yeah, agree. we will hash it with argon2 or bcrypt, it will be encrypted together with salt.

i think we can work on encryption when we work on user auth, which is another issue i have created. just tried to reduce the size of prs and focus on one at a time. but that's really good point and it's so critical we shouldn't miss.

how does the database structure look like to you guys, i tried but still i feel like maybe there should be more efficient way to do it? 🤔 .

we will probably change it anyway while we progress but i would like to try to make it as close as possible.

let me know if you guys have any questions or suggestions while examine this.

GrantMLambert commented:
We should also consider generating a random unique id string to each user and use that to identify them throughout the application as many apps do, as that will allow us to deal with situations in the future where it may be necessary to change a user's name, email or other information.

lugenx commented:
per your guys' feedback, i updated the structure of the models, and added some additional fields.

There are three models:
- User
- Habit (questions)
- Answer: includes references of habit id and user id.

Here are some examples:

User:
<img width="340" alt="image" src="https://user-images.githubusercontent.com/8403023/207227590-8eff0993-281a-40f1-841a-c50f59b409d6.png">

Habit
<img width="489" alt="image" src="https://user-images.githubusercontent.com/8403023/207227725-b54b7fb5-93c8-4f92-a34f-39d907885787.png">

Answer:
<img width="345" alt="image" src="https://user-images.githubusercontent.com/8403023/207227852-90ae51f1-da48-4e08-8aaa-47a7d5013b7d.png">

lugenx commented:
> We should also consider generating a random unique id string to each user and use that to identify them throughout the application as many apps do, as that will allow us to deal with situations in the future where it may be necessary to change a user's name, email or other information.

so because the database assigns each user a unique id, i think it shouldn't be a problem to change user's name, email etc.

AbhiBajpaishek commented:
LGTM!!

lugenx commented:
so i added the "roles" field to the user document in the mongoose schema in case we will have a different kinds of accounts in the future. see the example data below:

<img width="369" alt="image" src="https://user-images.githubusercontent.com/8403023/207714241-60d35246-214d-41c3-951c-ead8ad086f40.png">

i will merge this pr after a review, rest improvements can be made later.

lugenx commented:
@mannan0819 and @AbhiBajpaishek , i would like to see one of your approval before merging this. could you please review?

AbhiBajpaishek commented on server/src/controller/userController.js:
```diff
@@ -0,0 +1,29 @@
+};
+
+const userSignUp = async (req, res) => {
+  const name = req.body.name;
+  const email = req.body.email;
+  const postalCode = req.body.postalCode;
+  const password = req.body.password;
+  const roles = req.body.roles;
```
@lugenx

lugenx replied:
yeah, agree, much cleaner this way, let me update it.

lugenx replied:
this is done! @AbhiBajpaishek

mannan0819 reviewed (approved):
Looks good to me. Just need to keep in mind to not store password strings.
